design-proposals/kubevirt_ephemeral_csi_driver.md

```diff
@@ -0,0 +1,79 @@
+
+Another property of csi ephemeral volumes is that they can be specified inline in a pod spec and do not require an explicit persistent volume to function. When a pod using the csi ephemeral volume terminates, the csi driver is free to clean up the contents of the ephemeral volume, or perform any other operation it wants on the ephemeral volume.
+
+As said above a csi ephemeral volume is very close in function to an emptyDir, so why not simply use an emptyDir? The main motivation in suggesting an ephemeral csi driver is control over what happens when a pod exits. The following scenario can happen when using an emptyDir in combination with volume hotplug:
+
+1. We have a running VM, and have hotplugged one or more volumes into it. We have used an emptyDir as a stable path inside the virt-launcher which we used to attach the disks in those volumes to the VM (this is what the current volume hotplug implementation does).
+2. Something causes the virt-launcher to be force deleted (user action, resource pressure, whatever)
+3. When we normally terminate a VM, the virt-handler is notified we are going to terminate the VM, and is given a chance to unmount any volumes before the virt-launcher pod exits.
+4. When a pod that uses an emptyDir terminates, the kubelet will remove all the contents of the volume backing the emptyDir.
+5. Because we force terminated the virt-launcher pod, we are now in a race between the kubelet, and the virt-handler (normally the grace period will give virt-handler enough time). If the virt-handler is called to unmount and hotplugged volumes before the kubelet clears the emptyDir, we are fine. However if the kubelet is run before the virt-handler, it will blindly remove the contents of the emptyDir volume, which, in the hotplug volume case will be the disk.img files that have not been unmounted. Since it is a bind mount this will result in the disk.img file on the actual volumes being removed, and we have lost data.
+
+If we use a csi ephemeral volume controlled by a kubevirt specific csi driver, we can control the behavior in that scenario and unmount any mounted volumes inside the csi ephemeral volume before clearing it.
+
+## Goals
+
+* Design and implement a kubevirt specific csi driver that only supports **ephemeral volumes**.
+* Use csi ephemeral volumes to fully control what happens when virt-launcher pods terminate with regards to volume hotplug.
+
+## Non Goals
+
+Implement a full fledged CSI driver that also supports persistent volumes. We have other projects like the [hostpath csi driver](https://github.com/kubevirt/hostpath-provisioner) for that.
+
+
+## User Stories
+
+* As a KubeVirt user I want to be guaranteed that if I have hotplugged volumes in my VM, and the virt-launcher pod terminates unexpectedly, I will not lose any data on my volumes.
```

Comment:
out of curiosity, if the emptyDir's underlying medium is changed to something like "Memory" instead of the default which i think uses a disk directory, does the behavior of kubelet potentially deleting the disk file still occur?

Reply:
So I checked out the actual source for emptyDir, and if the medium if memory it just makes a tmpfs and uses that, but regardless of the medium we end up [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/emptydir/empty_dir.go#L484) which empties out the contents, and then below unmounts stuff based of the type of medium. I can quickly verify this by making a memory emptyDir for hotplug.

Reply:
bummer, alrighty, thanks for pointing out the emptrydir code. that is helpful